### PR TITLE
Feature/beta improvements

### PR DESCRIPTION
**Pull request by mpeyper**
**What**:

Fixed up some things:

1. Cleanup up some types
2. The type exports _should_ now by compatible the the pre-v4 DefinitelyTyped types ([discord comment](https://discord.com/channels/723559267868737556/723852521734668299/797598307068674048))
3. A potential bug that has existed since v3.6 when `result.all` was introduced

**Checklist**:

- [x] Tests
- [x] Ready to be merged

### CODE REVIEW

**Comment by mpeyper** [comment]
I'm not sure why the build has suddenly started caring about the lack of tests on `src/index.ts` and `src/pure.ts`... Probably the `kcd-scripts` changes?

Guess I'm looking at how we go about testing that next then.

File: src/core/index.ts [hunk 1]
```diff
@@ -1,19 +1,18 @@
...
   const resolvers: Array<() => void> = []
 
   const result: RenderResult<TValue> = {
     get all() {
-      return results.map(({ value, error }) => error ?? value)
+      return results.map(({ value, error }) => error ?? (value as TValue))
     },
     get current() {
-      const { value, error } = results[results.length - 1]
+      const { value, error } = results[results.length - 1] ?? {}
```
Comment (mpeyper) [inline src/core/index.ts:15]: This was the bug

File: src/__tests__/defaultRenderer.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    expect(actualRenderer).toEqual(expectedRenderer)
+  })
+
+  test('should resolve dom renderer as default renderer', () => {
+    jest.doMock('react-test-renderer', () => {
+      throw new Error('missing dependency')
+    })
+
+    const expectedRenderer = require('../dom/pure') as ReactHooksRenderer
+    const actualRenderer = require('..') as ReactHooksRenderer
```
Comment (mpeyper) [inline src/__tests__/defaultRenderer.test.ts:22]: had to use `require` in these tests otherwise the imports have resolved before the mocking occurs

Comment (joshuaellis) [reply src/__tests__/defaultRenderer.test.ts:22]: This is really interesting, I wasn't sure how to do this myself.

**Review by joshuaellis — APPROVED** [review]
Looks good to me.

**Pull request merged**